### PR TITLE
Added new commands

### PR DESCRIPTION
**Pull request by dlmr**
This PR adds a command to generate documentation for the available commands and the structure allows for generating quite complex documentation. The question is if we should use this instead of the manual documentation for the commands (that has not been updated in this PR). We should probably also look into making travis build and push this for us on every commit in all projects to avoid some manual work. Thoughts?

In general I think it might be a good idea after this has been merged to go over all of the documentation we have through the code to make sure that we use the features that we have available to us and that it's correct and easy to understand.

Another change is that we now handle non-empty directories a bit better. Now a prompt will ask the user what to do if the directory that it tries to use is not empty. Are the questions good here, and the language?

Also as can be seem we have added an alias to `init` named `new` that requires a argument that states the name of the folder to be created and in which the init should take place. This should make it easier for users to get started quickly and we support multiple mental models. The problem I see is that it might make it confusing with multiple commands that do the same thing. Thoughts about this? @sgulseth 
### How to test it

The changes done in the code can be tested by using the following commands:
- `init`
- `new`
- `markdown-commands`

In the case of `init` and `new` please try it with both empty and not empty directories.

An example of the generated documentation can be seen [here](https://gist.github.com/DLMR/5a72ec95560d58b6e243).
### Changes in the code
- Adds a new command "markdown-commands" that generate documentation for the available commands.
- Made non empty directory behavior better in "init".
- Created an alias for "init" named "new" that will always try to create a new folder.
- Changed slightly how the type information is generated for arrays and objects to work better with markdown.
- Some slight refactoring to work better with the new and updated commands.
- Made sure that we sort commands and options everywhere to be consistent.
- Updated incorrect and added new typedefs.
- Updated the documentation to reflect the changes.
- Updated the tests
- Updated ESDocs

### CODE REVIEW

**Comment by sgulseth** [comment]
I share your concern about adding different commands that in the end produces the same result. Ie. the user will type `roc init my-project` and roc would complain that my-project is not a template. I think at least we should "advertise" only one of them in the readme as the official way to start a new roc project, and perhaps let the other be a more of a hidden feature.

File: bin/index.js [hunk 1]
```diff
@@ -4,42 +4,80 @@ const runCli = require('../lib/cli').runCli;
...
+    name: 'force',
+    shortname: 'f',
+    validation: validators.isBoolean,
+    description: 'Ignore non empty directory warning.'
+}];
+
+const initArguments = [{
+    name: 'template',
+    validation: validators.isPath,
+    description: 'What template to use. Matches Github structure with Username/Repo.'
```
Comment (andreasrs) [inline bin/index.js]: s/What/The

File: bin/index.js [hunk 2]
```diff
@@ -4,42 +4,80 @@ const runCli = require('../lib/cli').runCli;
...
+}, {
+    name: 'version',
+    validation: validators.isString,
+    description: 'What version to use.'
```
Comment (andreasrs) [inline bin/index.js]: s/What/The

File: docs/config/commands.md [hunk 3]
```diff
@@ -39,6 +40,13 @@ parsedOptions       The parsed options given to the cli
 ### debug
 Debug will be set to `true` if `-d, --debug` was set. Should be used to print extra information when running the command. Otherwise it will be `false`.
 
+### info
+The same information object as `runCli` was invoked with, meaning it should have two properties.
+```
+version             The version of the running cli
+named               The name of the running cli
```
Comment (andreasrs) [inline docs/config/commands.md]: s/named/name ?

File: src/commands/init.js [hunk 4]
```diff
@@ -34,175 +34,227 @@ const templates = [{
...
+            /* eslint-disable no-process-exit */
+            process.exit(1);
+            /* eslint-enable */
+        }
+
+        toFetch = selectedTemplate.repo;
     }
 
-    return fetchTemplate(template, version);
+    return getVersions(toFetch)
```
Comment (andreasrs) [inline src/commands/init.js:61]: Would consider bringing some of these functions out with descriptive names.

File: src/commands/init.js [hunk 5]
```diff
@@ -34,175 +34,227 @@ const templates = [{
...
-
-            inquirer.prompt([{
-                type: 'rawlist',
-                name: 'option',
-                message: 'Selected a type',
-                choices: choices
-            }], answers => {
-                resolve(fetchTemplate(answers.option));
-            });
+function interativeMenu(directory, list) {
```
Comment (andreasrs) [inline src/commands/init.js]: s/interative/interactive

**Comment by andreasrs** [comment]
Overall looks good, have some comments that should be looked into.
- `roc init -l` (to list templates) or `roc-init web-react -l` (to list versions of web-react template) does not work in the way I expected it to. I don't have any luck with either of them. From looking at the code, it seems the 'list' option is not handled as expected in the functions that it is passed to.
- `roc markdown-commands` does not work outside of a project to document the "pure" roc commands like `init`.
- On command errors, `Command options problem A option was not valid.` - s/A/An

I'll finish the inline code comments in a while.

**Comment by dlmr** [comment]
@andreasrs We could list the templates using `roc init -l`however before we implement this I think we should thing about how we manage the default templates. I believe that this change could be done together with changing templates from being hard coded to be read from some backend.

The reason that `roc init web-react` does not work is since Roc expects that the first argument for the `init` command is the directory name, hence you can't list versions of it. :) 
_I was mixing it up with `new` here._

Good catch with `markdown-commands` not working as expected, will look for a fix to it.

Have you done the changes to the typos or should I do it as well?

**Comment by dlmr** [comment]
@sgulseth If we would promote one which would you go for? I assume `init`? We could also build in smarter error handling to mitigate problems that might happen because the user is mixing up the two commands.

**Comment by dlmr** [comment]
The commit above should fix:

> roc markdown-commands does not work outside of a project to document the "pure" roc commands like init.

**Pull request merged**